src/ripple/app/tx/impl/SetOracle.cpp

```diff
@@ -0,0 +1,259 @@
+        return ret;
+
+    if (ctx.tx.getFlags() & tfUniversalMask)
+        return temINVALID_FLAG;
+
+    auto const& dataSeries = ctx.tx.getFieldArray(sfPriceDataSeries);
+    if (dataSeries.size() == 0 || dataSeries.size() > maxOracleDataSeries)
+        return temBAD_ARRAY_SIZE;
+
+    auto invalidLength = [&](auto const& sField, std::size_t length) {
+        return ctx.tx.isFieldPresent(sField) &&
+            ctx.tx[sField].length() > length;
+    };
+
+    if (invalidLength(sfProvider, maxOracleProvider) ||
+        invalidLength(sfURI, maxOracleURI) ||
+        invalidLength(sfAssetClass, maxOracleSymbolClass))
+        return temMALFORMED;
+
+    return preflight2(ctx);
+}
+
+TER
+SetOracle::preclaim(PreclaimContext const& ctx)
+{
+    auto const sleSetter =
+        ctx.view.read(keylet::account(ctx.tx.getAccountID(sfAccount)));
+    if (!sleSetter)
+        return terNO_ACCOUNT;
+
+    // lastUpdateTime must be within maxLastUpdateTimeDelta seconds
+    // of the last closed ledger
+    using namespace std::chrono;
+    std::size_t const closeTime =
+        duration_cast<seconds>(ctx.view.info().closeTime.time_since_epoch())
+            .count();
+    std::size_t const lastUpdateTime = ctx.tx[sfLastUpdateTime];
+    if (lastUpdateTime <= epoch_offset.count())
+        return tecINVALID_UPDATE_TIME;
+    std::size_t const lastUpdateTimeEpoch =
+        lastUpdateTime - epoch_offset.count();
+    if (lastUpdateTimeEpoch < closeTime ||
+        lastUpdateTimeEpoch > (closeTime + maxLastUpdateTimeDelta))
+        return tecINVALID_UPDATE_TIME;
+
+    hash_set<uint256> pairs;
+    for (auto const& entry : ctx.tx.getFieldArray(sfPriceDataSeries))
+    {
+        if (!entry.isFieldPresent(sfAssetPrice))
+            return temMALFORMED;
+        auto const hash = tokenPairHash(entry);
+        if (pairs.contains(hash))
+            return tecDUPLICATE;
+        pairs.emplace(hash);
+    }
+
+    if (auto const sle = ctx.view.read(keylet::oracle(
+            ctx.tx.getAccountID(sfAccount), ctx.tx[sfOracleDocumentID])))
+    {
+        // update
+
+        // lastUpdateTime must be more recent than the previous one
+        if (ctx.tx[sfLastUpdateTime] <= (*sle)[sfLastUpdateTime])
+            return tecINVALID_UPDATE_TIME;
+
+        if (ctx.tx[sfAccount] != sle->getAccountID(sfOwner))
+            return tecNO_PERMISSION;
+
+        if (ctx.tx.isFieldPresent(sfProvider) ||
+            ctx.tx.isFieldPresent(sfAssetClass))
+            return temMALFORMED;
+
+        for (auto const& entry : sle->getFieldArray(sfPriceDataSeries))
+        {
+            auto const hash = tokenPairHash(entry);
+            if (!pairs.contains(hash))
+                pairs.emplace(hash);
+        }
+    }
+    else
+    {
+        // create
+
+        if (!ctx.tx.isFieldPresent(sfProvider) ||
+            !ctx.tx.isFieldPresent(sfAssetClass))
+            return temMALFORMED;
+    }
+
+    if (pairs.size() > maxOracleDataSeries)
```

Comment:
Yes, it's necessary. There are two use-cases. The preflight checks if the transaction's `PriceDataSeries` array exceeds ten.  Even if `PriceDataSeries` contains just one element in the transaction, it's possible that the `PriceOracle` object already has ten elements in `PriceDataSeries`. The second check confirms that the `PriceDataSeries` is less or equal to ten after the update.

Reply:
Why is this check not just done in `doApply`? Seems easier and more efficient that way, to avoid replicating work.
The way `preclaim` vs `doApply` were explained to me is that `preclaim` does as many checks as is reasonable to do with read-only access to the ledger, but checks can also be done in `doApply` so it's often not worth it to do checks in `preclaim` that are easier to just perform in `doApply`.

Reply:
There is a lot going on with the error checking. Especially now that we also added the delete. So I'd rather separate as much as possible of the error handling from the apply handling. Apply is pretty clean now and doesn't have any transaction parameters related error checking.